express-api/src/services/properties/propertiesServices.ts

```diff
@@ -110,6 +125,395 @@ const getPropertiesForMap = async (filter?: MapFilter) => {
+    UpdatedOn: existentBuilding ? new Date() : undefined,
+    CreatedOn: existentBuilding ? undefined : new Date(),
+    Location: {
+      x: row.Longitude,
+      y: row.Latitude,
+    },
+    AdministrativeAreaId: adminAreaId,
+    IsSensitive: false,
+    IsVisibleToOtherAgencies: true,
+    PropertyTypeId: 0,
+    RentableArea: 0,
+    BuildingTenancy: '123',
+    BuildingFloorCount: 0,
+    TotalArea: 0,
+    Evaluations: currRowEvaluations,
+    Fiscals: currRowFiscals,
+  };
+};
+
+type Lookups = {
+  classifications: PropertyClassification[];
+  constructionTypes: BuildingConstructionType[];
+  predominateUses: BuildingPredominateUse[];
+  agencies: Agency[];
+  adminAreas: AdministrativeArea[];
+  userAgencies: number[];
+};
+
+type BulkUploadRowResult = {
+  action: 'inserted' | 'updated' | 'ignored' | 'error';
+  reason?: string;
+};
+
+const importPropertiesAsJSON = async (
+  worksheet: WorkSheet,
+  user: User,
+  roles: string[],
+  resultId: number,
+) => {
+  const sheetObj: Record<string, any>[] = xlsx.utils.sheet_to_json(worksheet);
+  const classifications = await AppDataSource.getRepository(PropertyClassification).find({
+    select: { Name: true, Id: true },
+  });
+  const constructionTypes = await AppDataSource.getRepository(BuildingConstructionType).find({
+    select: { Name: true, Id: true },
+  });
+  const predominateUses = await AppDataSource.getRepository(BuildingPredominateUse).find({
+    select: { Name: true, Id: true },
+  });
+  const agencies = await AppDataSource.getRepository(Agency).find({
+    select: { Name: true, Id: true, Code: true },
+  });
+  const adminAreas = await AppDataSource.getRepository(AdministrativeArea).find({
+    select: { Name: true, Id: true },
+  });
+  const userAgencies = await userServices.getAgencies(user.Username);
+  const lookups: Lookups = {
+    classifications,
+    constructionTypes,
+    predominateUses,
+    agencies,
+    adminAreas,
+    userAgencies,
+  };
+  const results: Array<BulkUploadRowResult> = [];
+  let queuedParcels = [];
+  let queuedBuildings = [];
+  const queryRunner = AppDataSource.createQueryRunner();
+  await queryRunner.startTransaction();
+  try {
+    for (let rowNum = 0; rowNum < sheetObj.length; rowNum++) {
+      const row = sheetObj[rowNum];
+      if (row.PropertyType === undefined) {
+        results.push({ action: 'ignored', reason: 'Must specify PropertyType for this row.' });
+        continue;
+      }
+      if (row.PropertyType === 'Land') {
+        const existentParcel = await queryRunner.manager.findOne(Parcel, {
+          where: { PID: numberOrNull(row.PID) },
+        });
+        try {
+          const parcelToUpsert = await makeParcelUpsertObject(
+            row,
+            user,
+            roles,
+            lookups,
+            queryRunner,
+            existentParcel,
+          );
+          queuedParcels.push(parcelToUpsert);
+          results.push({ action: existentParcel ? 'updated' : 'inserted' });
```

Comment:
If we push this success action here, could it still fail when the queuedParcels/Buildings goes to insert?
I'm thinking this might return false positives.

Reply:
Well in theory, anything that would cause the parcelToUpsert object to be invalid would have already thrown an error inside the makeParcelUpsertObject function. But that's a good point, the insert could still fail for some unexpected reason when it actually makes the save call. I suppose this may be another reason the old API does everything one at a time, but in the ideal case where all rows from the spreadsheet contain new PIDs it's much faster to do batched insert. I'll have to think about this a bit.

Reply:
I definitely like the batch better. Just wondering if there's a way to see which items failed in that batch insert, then update the results maybe.